simple/app_operator.go

```diff
@@ -0,0 +1,222 @@
+package simple
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/grafana/grafana-app-sdk/app"
+	"github.com/grafana/grafana-app-sdk/k8s"
+	"github.com/grafana/grafana-app-sdk/metrics"
+	"github.com/grafana/grafana-app-sdk/operator"
+	"github.com/grafana/grafana-app-sdk/resource"
+)
+
+// StandaloneOperator runs an app as a standalone operator, capable of exposing admission (validation, mutation)
+// and conversion as webhooks, and running a main control loop with reconcilers and watchers.
+// It relies on the Kinds managed by the app.App already existing in the API server it talks to, either as CRD's
+// or another type. It does not support certain advanced app.App functionality which is not natively supported by
+// CRDs, such as arbitrary subresources (app.App.CallSubresource). It should be instantiated with NewStandaloneOperator.
+type StandaloneOperator struct {
+	provider app.Provider
+}
+
+// NewStandaloneOperator creates a new, properly-initialized instance of a StandaloneOperator,
+// which will use the given app.Provider to instantiate a new underlying app.
+func NewStandaloneOperator(provider app.Provider) (*StandaloneOperator, error) {
+	if provider == nil {
+		return nil, errors.New("provider cannot be nil")
+	}
+	return &StandaloneOperator{
+		provider: provider,
+	}, nil
+}
+
+type OperatorAppConfig struct {
+	// WebhookConfig contains configuration information for exposing k8s webhooks.
+	// This can be empty if your App does not implement ValidatorApp, MutatorApp, or ConversionApp
+	WebhookConfig OperatorWebhookConfig
+	// MetricsConfig contains the configuration for exposing prometheus metrics, if desired
+	MetricsConfig MetricsConfig
+	// TracingConfig contains the configuration for sending traces, if desired
+	TracingConfig TracingConfig
+	// AppConfig contains the configuration needed for creating and running the underlying App
+	AppConfig app.Config
+	// Filesystem is an fs.FS that can be used in lieu of the OS filesystem.
+	// if empty, it defaults to os.DirFS(".")
+	Filesystem fs.FS
+}
+
+type OperatorWebhookConfig struct {
+	// Port is the port to open the webhook server on
+	Port int
+	// TLSConfig is the TLS Cert and Key to use for the HTTPS endpoints exposed for webhooks
+	TLSConfig k8s.TLSConfig
+}
+
+type capabilities struct {
+	conversion bool
+	mutation   bool
+	validation bool
+}
+
+// Run runs the StandaloneOperator using the provided config, until stopCh is closed or receives a message, or an unrecoverable error occurs
+func (s *StandaloneOperator) Run(config OperatorAppConfig, stopCh <-chan struct{}) error {
```

Comment:
SUGGESTION – how about inverting the order of execution and requiring the confing in `NewStandaloneOperator` but passing the provider here? That would allow one to run multiple apps with one operator.

Also, we could switch to context instead of using channels.

```suggestion
func (s *StandaloneOperator) Run(ctx context.Context, provider app.Provider) error {
```

Reply:
With how it currently works under the hood without rewriting substantial parts of code, you can't run multiple apps under the same StandalongOperator, because `Run` sets up and runs an operator that includes a `k8s.WebhookServer`, which is bound to a specific port from the config. If you tried to call `Run` again with a different (or event the same) provider, it would try to re-bind to that port (based on the config) and fail. We could have the config _and_ the provider passed to `Run`, or perhaps a "minimal" config, but at that point, it seems like it would be just as easy for a user to have two `StandalongOperator` instances.

I have converted `Run` to use a `context.Context`, though, to match `app.Runnable`.

Reply:
How about future-proofing the API? I think we should consider rewriting some code to make it possible to `Run` multiple apps under one operator, which would include having a shared webhook server listener, etc. etc.

Reply:
OK, I didn't want to touch the underlying `k8s` code due to that breaking extant projects, and what we needed anyway was the ability to have `Run` use the same webserver, so I introduced `app.SingletonRunner`, which has a wrapped `Runnable` that multiple calls to `Run` will only run the underlying `Run` if it isn't already running (started this as specific logic in `simple.StandaloneOperator`, but it felt like it might be more useful as generic logic for any author to use). I think we'll eventually need to re-write/move/update the `k8s.WebhookServer` logic, but for now it seemed most expedient to re-use it with a little bit of a hack underneath. This ensures that he API is good, even if we need to update the underlying code later.

Reply:
Nice, thank you!